src/components/shapes/draw.js

```diff
@@ -579,6 +599,208 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
     }
 }
 
+function drawLabel(gd, index, options, shapeGroup) {
+    // Remove existing label
+    shapeGroup.selectAll('.shape-label').remove();
+
+    // If no label, return
+    if(!options.label) return;
+
+    var labelGroupAttrs = {
+        'data-index': index,
+    };
+    var text = options.label.text;
+    var font = options.label.font;
+
+    var labelTextAttrs = {
+        'data-notex': 1
+    };
+
+    var labelGroup = shapeGroup.append('g')
+        .attr(labelGroupAttrs)
+        .classed('shape-label', true);
+    var labelText = labelGroup.append('text')
+        .attr(labelTextAttrs)
+        .classed('shape-label-text', true)
+        .text(text);
+
+    // If x0, x1, y0, y1 are defined explicitly, use those values
+    // Otherwise, use shape bounding box
+    var shapex0, shapex1, shapey0, shapey1;
```

Comment:
Right now, the positioning on reversed axes is not correct e.g. when `autorange: 'reversed'` is used.
To fix that perhaps you need to adjust this block to something like:
```js
    // If x0, x1, y0, y1 are defined explicitly, use those values
    // Otherwise, use shape bounding box
    var _shapex0, _shapex1, _shapey0, _shapey1;
    if('x0' in options && 'x1' in options && 'y0' in options && 'y1' in options) {
        // Setup conversion functions
        var xa = Axes.getFromId(gd, options.xref);
        var xRefType = Axes.getRefType(options.xref);
        var ya = Axes.getFromId(gd, options.yref);
        var yRefType = Axes.getRefType(options.yref);
        var x2p = helpers.getDataToPixel(gd, xa, false, xRefType);
        var y2p = helpers.getDataToPixel(gd, ya, true, yRefType);

        _shapex0 = x2p(options.x0);
        _shapex1 = x2p(options.x1);
        _shapey0 = y2p(options.y0);
        _shapey1 = y2p(options.y1);
    } else {
        // Get shape bounding box
        var shapeBBox = shapeGroup.selectAll('path').node().getBoundingClientRect();

        _shapex0 = shapeBBox.left;
        _shapex1 = shapeBBox.right;
        _shapey0 = shapeBBox.top;
        _shapey1 = shapeBBox.bottom;
    }

    var shapex0 = Math.min(_shapex0, _shapex1);
    var shapex1 = Math.max(_shapex0, _shapex1);
    var shapey0 = Math.min(_shapey0, _shapey1);
    var shapey1 = Math.max(_shapey0, _shapey1);
```

Reply:
@archmoj The positioning for reversed axes seems fine to me -- I am taking the `min` and `max` later, inside the `calcTextPosition()` function. If you are seeing cases where the position is wrong, could you give an example?

Reply:
Resolved.